concurrency/SerDeUtils.java

```diff
@@ -0,0 +1,456 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package concurrency;
+
+import org.apache.hadoop.hive.common.JavaUtils;
+import org.apache.hadoop.hive.serde2.objectinspector.*;
+import org.apache.hadoop.hive.serde2.objectinspector.primitive.*;
+import org.apache.hadoop.io.BytesWritable;
+import org.apache.hadoop.io.Text;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * SerDeUtils.
+ *
+ */
+public final class SerDeUtils {
+
+  public static final char QUOTE = '"';
+  public static final char COLON = ':';
+  public static final char COMMA = ',';
+  public static final String LBRACKET = "[";
+  public static final String RBRACKET = "]";
+  public static final String LBRACE = "{";
+  public static final String RBRACE = "}";
+
+  private static ConcurrentHashMap<String, Class<?>> serdes =
+    new ConcurrentHashMap<String, Class<?>>();
+
+  public static void registerSerDe(String name, Class<?> serde) {
+    if (serdes.containsKey(name)) {
+      throw new RuntimeException("double registering serde " + name);
+    }
+    serdes.put(name, serde);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **50 and 53**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 If the key already exists, the code should throw an error at line: **51**. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.